File: RemixedDungeon/src/main/java/com/watabou/pixeldungeon/actors/buffs/Poison.java

package com.watabou.pixeldungeon.actors.buffs;

import com.nyrds.pixeldungeon.ml.R;
import com.nyrds.platform.util.StringsManager;
import com.watabou.pixeldungeon.Badges;
import com.watabou.pixeldungeon.Dungeon;
import com.watabou.pixeldungeon.ResultDescriptions;
import com.watabou.pixeldungeon.actors.Char;
import com.watabou.pixeldungeon.actors.hero.Doom;
import com.watabou.pixeldungeon.effects.CellEmitter;
import com.watabou.pixeldungeon.effects.particles.PoisonParticle;
import com.watabou.pixeldungeon.items.rings.RingOfElements.Resistance;
import com.watabou.pixeldungeon.sprites.CharSprite;
import com.watabou.pixeldungeon.ui.BuffIndicator;
import com.watabou.pixeldungeon.utils.GLog;
import com.watabou.pixeldungeon.utils.Utils;

public class Poison extends Buff implements Doom {

	@Override
	public int icon() {
		return BuffIndicator.POISON;
	}

	@Override
	public void charAct() {
		float timeLeft = cooldown();
		target.damage( (int)(timeLeft / 3) + 1, this );
	}

	@Override
	public boolean act() {
		detach();
		return true;
	}

	public static float durationFactor(Char ch ) {
		if(ch==null) { //primary to mask bug in Remixed Additions
			return 1;
		}

		Resistance r = ch.buff( Resistance.class );
		return r != null ? r.durationFactor() : 1;
	}

	@Override
	public void onHeroDeath() {
		Badges.validateDeathFromPoison();
		
		Dungeon.fail( Utils.format( ResultDescriptions.getDescription(ResultDescriptions.Reason.POISON), Dungeon.depth ) );
        GLog.n(StringsManager.getVar(R.string.Poison_Death));
	}

	@Override
	public void attachVisual() {
		CellEmitter.center(target.getPos()).burst(PoisonParticle.SPLASH, 5);
        target.showStatus(CharSprite.NEGATIVE, StringsManager.getVar(R.string.Char_StaPoisoned));
	}
}

File: scripts/buffs/Poison.lua
---
--- Poison: DoT ticks scale with remaining duration (was actors/buffs/Poison.java)
---
local RPD  = require "scripts/lib/commonClasses"
local buff = require "scripts/lib/buff"

return buff.init{
    desc  = function ()
        return {
            icon          = 3, -- BuffIndicator.POISON
            name          = "PoisonBuff_Name",
            info          = "PoisonBuff_Info",
        }
    end,

    charAct = function(self, buff)
        local timeLeft = buff:cooldown()
        buff.target:damage(math.floor(timeLeft / 3) + 1, buff)
    end,

    act = function(self, buff)
        buff:detach()
    end,

    onHeroDeath = function(self, buff)
        RPD.Badges:validateDeathFromPoison()
        RPD.Dungeon:fail(RPD.JavaUtils:format(
            RPD.ResultDescriptions:getDescription(RPD.ResultReason.POISON), { RPD.Dungeon.depth }))
        RPD.glogn(RPD.textById("Poison_Death"))
    end,

    attachVisual = function(self, buff)
        RPD.Sfx.CellEmitter:center(buff.target:getPos()):burst(RPD.Sfx.PoisonParticle.SPLASH, 5)
        local CharSprite = luajava.bindClass("com.watabou.pixeldungeon.sprites.CharSprite")
        buff.target:showStatus(CharSprite.NEGATIVE, RPD.textById("Char_StaPoisoned"))
    end
}
